### PR TITLE
refactor: take advantage of structured clone algorithm in the remote module

### PR DESCRIPTION
**Pull request by miniak**
#### Description of Change
Take advantage of structured clone algorithm based IPC introduced by #20214:
- `Date`, `ArrayBuffer` and `ArrayBufferView` can now be serialized as values.
- Keeping special handling to preserve `Buffer`, otherwise it will be converted to `UInt8Array`.
- Use native serialization for `Boolean`, `Number`, `String` and `RegExp`.
- `Error` is now serialized natively. Preserve custom properties in the `remote` module.

#### Checklist
- [x] PR description included and stakeholders cc'd
- [x] `npm test` passes
- [x] PR title follows semantic [commit guidelines](https://github.com/electron/electron/blob/master/docs/development/pull-requests.md#commit-message-guidelines)
- [x] [PR release notes](https://github.com/electron/clerk/blob/master/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/master/README.md#examples).

#### Release Notes
Notes: The `remote` module now properly serializes `Boolean`, `Number`, `String` and `RegExp` instances.

### CODE REVIEW

**Comment by miniak** [comment]
@nornagon can you please review?

File: lib/browser/remote/server.ts [hunk 1]
```diff
@@ -186,8 +182,6 @@ const valueToMeta = function (sender: electron.WebContents, contextId: string, v
       value: value.name
     })
     return { type, members }
-  } else if (type === 'date') {
-    return { type, value: value.getTime() }
   } else {
     return {
       type: 'value',
```
[the remark this answers is not in the document]
Comment (miniak) [reply lib/browser/remote/server.ts]: I know, but the default serialization looses custom properties

[the remark this answers is not in the document]
Comment (miniak) [reply lib/browser/remote/server.ts]: There is a test explicitly enforcing that, which started failing when I did the change for `Error` as well

File: lib/browser/remote/server.ts [hunk 2]
```diff
@@ -128,14 +124,14 @@ const valueToMeta = function (sender: electron.WebContents, contextId: string, v
     // Recognize certain types of objects.
     if (value === null) {
```
[the remark this answers is not in the document]
Comment (miniak) [reply lib/browser/remote/server.ts]: moved the check for `null` into the `isSerializable`, and removed from `lib/renderer/api/remote.js` as well

File: lib/browser/remote/server.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (miniak) [reply lib/browser/remote/server.ts]: done + added code to preserve custom properties

**Comment by miniak** [comment]
@nornagon do we mind loosing the `from` property indicating which process the `Error` came from, which was added by the old `errorUtils` serializer?

File: lib/common/remote/type-utils.ts [hunk 3]
```diff
@@ -10,3 +10,16 @@ export function isPromise (val: any) {
...
   )
 }
+
+const serializableTypes = [
+  Boolean,
+  Number,
+  String,
+  Date,
+  RegExp,
+  ArrayBuffer
```
Comment (nornagon) [inline lib/common/remote/type-utils.ts]: `Map`, `Set`?

File: lib/renderer/api/remote.js [hunk 4]
```diff
@@ -43,17 +41,17 @@ function wrapArgs (args, visited = new Set()) {
       }
       visited.delete(value)
```
Comment (nornagon) [inline lib/renderer/api/remote.js]: it's still a mystery to me why we have two separate serialization algorithms

File: lib/common/remote/type-utils.ts (hunk 3 above)
Comment (miniak) [reply lib/common/remote/type-utils.ts]: those can contain other objects, which require custom serialization.

File: lib/renderer/api/remote.js (hunk 4 above)
Comment (miniak) [reply lib/renderer/api/remote.js]: do you mean main -> renderer and renderer -> main?

Comment (nornagon) [reply lib/renderer/api/remote.js]: yes

**Pull request merged**